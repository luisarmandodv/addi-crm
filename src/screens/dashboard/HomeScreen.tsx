import React, { useEffect, useState } from 'react';
import {
  View,
  StyleSheet,
  StatusBar,
  ActivityIndicator,
  ScrollView,
  SafeAreaView,
  SectionList,
  TouchableOpacity,
} from 'react-native';
import { Avatar, Title, Caption, Text } from 'react-native-paper';
import { useNavigation } from '@react-navigation/native';
import { useGetUsers } from '@hooks/useGetUsers';
import Modal from '@components/Modal';
import useClientsStore from '@stores/useClientsStore';
import useModal from '@stores/useModal';
import { colors } from '@theme/colors';
import { formStyles, typography } from '@theme/globalStyles';
import { UserData } from 'src/types';

export function HomeScreen() {
  const setIsModalVisible = useModal(
    (state: { setIsModalVisible: (value: boolean) => void }) =>
      state.setIsModalVisible
  );
  const { leads, prospects } = useClientsStore(({ leads, prospects }) => ({
    leads,
    prospects,
  }));
  const { data, isLoading } = useGetUsers();
  const { navigate }: any = useNavigation();

  const Loader = React.useMemo(() => {
    return (
      <ActivityIndicator
        testID="loader"
        hidesWhenStopped
        size="large"
        animating={isLoading}
        style={styles.activityIndicator}
      />
    );
  }, [isLoading]);

  const list = React.useMemo(() => {
    const sectionList = []
    
    if (leads && leads.length) {
      sectionList.push({
        title: 'Leads',
        data: leads as UserData[],
      })
    }

    if (prospects && prospects.length) {
      sectionList.push({
        title: 'Prospect',
        data: prospects as UserData[],
      });
    }

    return sectionList;
  }, [leads, prospects]);

  const List = React.useMemo(() => {
    return (
      <SectionList
        sections={list}
        keyExtractor={(item, index) => item.firstName + index.toString()}
        renderSectionHeader={({ section: { title } }) => (
          <Title style={styles.header}>{title}</Title>
        )}
        ListEmptyComponent={<View><Text>Empty</Text></View>}
        renderItem={({ item, section, index }) => {
          return (
            <TouchableOpacity
              testID={`button`}
              onPress={() => {
                navigate('UserDetails', { user: item, section: section.title });
              }}
            >
              <View style={styles.itemContainer}>
                <View style={{ flexDirection: 'row', alignItems: 'center' }}>
                  <View style={{ marginHorizontal: 15 }}>
                    <Avatar.Image
                      style={{
                        height: 50,
                        width: 50,
                        borderRadius: 50 / 2,
                        alignSelf: 'center',
                      }}
                      source={{
                        uri: item.image,
                      }}
                      size={50}
                    />
                  </View>
                  <View>
                    <Title style={typography.paragraph}>
                      {item.firstName + ' ' + item.lastName}
                    </Title>
                    <Caption style={{ color: colors.gray }}>
                      {item.email}
                    </Caption>
                    {section.title === 'Leads' ? (
                      <Text style={{ color: colors.warning }}>In Progress</Text>
                    ) : (
                      <Text style={{ color: colors.green }}>Approved</Text>
                    )}
                  </View>
                </View>
              </View>
            </TouchableOpacity>
          );
        }}
      />
    );
  }, [list]);

  return (
    <SafeAreaView style={styles.container}>
      <StatusBar barStyle="dark-content" />
      {Loader}
      <Modal onLeftTextPress={() => setIsModalVisible(false)} />
      <ScrollView>{List}</ScrollView>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: { flex: 1, backgroundColor: colors.primary },
  activityIndicator: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  },
  itemContainer: {
    backgroundColor: colors.brown,
    marginBottom: 10,
    padding: 10,
    margin: 8,
    borderRadius: 10,
  },
  header: {
    marginLeft: 10,
    fontSize: 20,
    fontWeight: 'bold',
  },
});
