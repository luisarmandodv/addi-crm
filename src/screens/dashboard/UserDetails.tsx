import React, { useCallback, useMemo } from 'react';
import {
  View,
  StyleSheet,
  SafeAreaView,
  ScrollView,
  ActivityIndicator,
} from 'react-native';
import { Avatar, Title, Caption, Text, Button } from 'react-native-paper';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import type { NativeStackScreenProps } from '@react-navigation/native-stack';
import { useNavigation } from '@react-navigation/native';
import { colors } from '@theme/colors';
import { useGetNRI } from '@hooks/useGetNRI';
import { useJudicialRecords } from '@hooks/useJudicialRecords';
import { useProspectQualification } from '@hooks/useProspectQualification';
import { AuthNavigatorRoutesprops } from '@routes/DashboardRoutes';
import useClientsStore from '@stores/useClientsStore';
import {
  setNotificationHandler,
  scheduleNotificationAsync,
} from 'expo-notifications';
import { UserData } from 'src/types';

interface ProfileProps {
  user: UserData;
  section: string;
}

type Props = NativeStackScreenProps<AuthNavigatorRoutesprops>;

setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: false,
    shouldSetBadge: false,
  }),
});

function UserInfo({ user }: { user: UserData }) {
  return (
    <View style={[styles.userInfoSection, styles.userInfoContainer]}>
      <View style={styles.userInfoRow}>
        <Avatar.Image source={{ uri: user.image }} size={80} />
        <View style={styles.userInfoText}>
          <Title
            style={[styles.title, styles.userInfoTitle]}
          >{`${user.firstName} ${user.lastName}`}</Title>
          <Caption style={styles.caption}>{user.username}</Caption>
        </View>
      </View>
      <View style={[styles.userInfoSection, styles.userInfoDetails]}>
        <View style={styles.row}>
          <Icon name="map-marker-radius" color="#777777" size={20} />
          <Text style={styles.loadingText}>{user.address.city}</Text>
        </View>
        <View style={styles.row}>
          <Icon name="phone" color="#777777" size={20} />
          <Text style={styles.loadingText}>{user.phone}</Text>
        </View>
        <View style={styles.row}>
          <Icon name="email" color="#777777" size={20} />
          <Text style={styles.loadingText}>{user.email}</Text>
        </View>
      </View>
    </View>
  );
}

function InfoBox({
  title,
  isLoading,
  data,
  isLeadsUser,
}: {
  title: string;
  isLoading: boolean;
  data: any;
  isLeadsUser: boolean;
}) {
  const Result = useMemo(() => {
    if (isLoading) {
      return (
        <ActivityIndicator
          testID={`${title}-loader`}
          hidesWhenStopped
          size="small"
          animating={true}
        />
      );
    }
    if (data?.data?.meetsRequirements || !isLeadsUser) {
      return <Icon name="check" color={colors.green} size={40} />;
    }
    return <Icon name="close" color={colors.danger} size={40} />;
  }, [isLoading, data, isLeadsUser]);

  return (
    <View style={styles.infoBox}>
      <Caption style={styles.infoBoxText}>{title}</Caption>
      {Result}
    </View>
  );
}

function InternalQualification({ user }: { user: UserData }) {
  const qualification = useProspectQualification();
  const percentage =
    user.qualification ||
    qualification.data?.qualification as number
  const isPropspect = Boolean(percentage > 60);
  const color = isPropspect ? colors.green : colors.danger;
  const setProspects = useClientsStore((state) => state.setProspects);
  const navigation = useNavigation();
  const result = useMemo(() => {
    if (qualification.isFetching || qualification.isLoading) {
      return (
        <ActivityIndicator
          testID={`internal-qualification-loader`}
          hidesWhenStopped
          size="small"
          animating={true}
        />
      );
    }
    return (
      <Text
        testID="internal-qualification-text"
        style={[styles.internalQualificationText, { color }]}
      >
        {percentage}/100
      </Text>
    );
  }, [qualification, color, percentage]);

  const onPress = useCallback(() => {
    const schedulingOptions = {
      content: {
        title: 'New Prospect Added',
        body: ``,
        sound: true,
        color: 'blue',
      },
      trigger: {
        seconds: 1,
      },
    };

    scheduleNotificationAsync(schedulingOptions);

    setProspects({ ...user, qualification: percentage });
    navigation.goBack();
  }, [user, percentage]);

  return (
    <>
      <View style={styles.leadsContainer}>
        <Title
          testID="internal-qualification-title"
          style={[styles.title, styles.userInfoTitle]}
        >
          Internal Qualification
        </Title>
        {result}
      </View>
      {!user.qualification && <Button
        mode="contained"
        disabled={!isPropspect}
        onPress={onPress}
        style={{ marginTop: 50, marginHorizontal: 30 }}
      >
        Add new prospect
      </Button>}
    </>
  );
}

export function UserDetails({ route }: Props) {
  const { user, section } = route.params as ProfileProps;
  const isLeadsUser = Boolean(section === 'Leads');
  const NRI = useGetNRI(isLeadsUser);
  const juficialRecords = useJudicialRecords(isLeadsUser);

  const renderInternalQualification = useMemo(() => {
    if (NRI.isFetching && juficialRecords.isFetching) return;
    if (
      !isLeadsUser ||
      (NRI.data?.meetsRequirements && juficialRecords.data?.meetsRequirements)
    ) {
      return <InternalQualification user={user} />;
    }
  }, [NRI, juficialRecords, user, isLeadsUser]);

  return (
    <SafeAreaView style={styles.container}>
      <ScrollView>
        <UserInfo user={user} />
        <View style={styles.infoBoxWrapper}>
          <InfoBox
            title="National Registry Identification"
            isLoading={NRI.isLoading || NRI.isFetching}
            data={NRI}
            isLeadsUser={isLeadsUser}
          />
          <InfoBox
            title="Judicial Records"
            isLoading={juficialRecords.isLoading || juficialRecords.isFetching}
            data={juficialRecords}
            isLeadsUser={isLeadsUser}
          />
        </View>
        {renderInternalQualification}
      </ScrollView>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.primary,
  },
  userInfoSection: {
    paddingHorizontal: 30,
    marginBottom: 25,
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
  },
  caption: {
    fontSize: 14,
    lineHeight: 14,
    fontWeight: '500',
  },
  row: {
    flexDirection: 'row',
    marginBottom: 16,
  },
  infoBoxWrapper: {
    margin: 10,
    padding: 10,
    backgroundColor: colors.brown,
    borderRadius: 10,
    flexDirection: 'row',
    height: 100,
  },
  infoBox: {
    width: '50%',
    alignItems: 'center',
    justifyContent: 'center',
  },
  infoBoxText: {
    marginBottom: 10,
  },
  loadingText: {
    color: '#777777',
  },
  internalQualificationText: {
    fontSize: 25,
    alignSelf: 'center',
    marginTop: 10,
  },
  userInfoContainer: {
    backgroundColor: colors.brown,
    marginBottom: 10,
    padding: 10,
    margin: 8,
    borderRadius: 10,
  },
  leadsContainer: {
    justifyContent: 'center',
    alignItems: 'center',
    margin: 10,
    padding: 10,
    borderRadius: 10,
    backgroundColor: colors.brown,
  },
  userInfoRow: { flexDirection: 'row', marginLeft: 7 },
  userInfoText: { marginLeft: 20 },
  userInfoTitle: { marginTop: 15, marginBottom: 5 },
  userInfoDetails: { marginTop: '10%', marginLeft: '-10%' },
});
