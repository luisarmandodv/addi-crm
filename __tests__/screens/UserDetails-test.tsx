import React from 'react';
import { render, screen } from '@testing-library/react-native';
import { UserDetails } from '@screens/dashboard/UserDetails';
import { QueryClient, QueryClientProvider } from 'react-query';

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));
jest.mock('@hooks/useGetNRI', () => ({
  useGetNRI: () => ({
    data: {},
  }),
}));

jest.mock('react-native-vector-icons/MaterialCommunityIcons', () => 'MockedMaterialCommunityIcons');

describe('UserDetails Component', () => {
  const client = new QueryClient();

  it('Should render correctly component', () => {
    const user = {
      image: 'user-image-url',
      firstName: 'John',
      lastName: 'Doe',
      username: 'johndoe123',
      address: { city: 'New York' },
      phone: '123-456-7890',
      email: 'johndoe@example.com',
    };

    const section = 'Leads';
    render(
      <QueryClientProvider client={client}>
        <UserDetails route={{ params: { user, section } }} />
      </QueryClientProvider>
    );
  });

  it('Should render snapshot', () => {
    const user = {
      image: 'user-image-url',
      firstName: 'John',
      lastName: 'Doe',
      username: 'johndoe123',
      address: { city: 'New York' },
      phone: '123-456-7890',
      email: 'johndoe@example.com',
    };

    const section = 'Leads';
    render(
      <QueryClientProvider client={client}>
        <UserDetails route={{ params: { user, section } }} />
      </QueryClientProvider>
    );

    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('renders correctly with user data', () => {
    const user = {
      image: 'user-image-url',
      firstName: 'John',
      lastName: 'Doe',
      username: 'johndoe123',
      address: { city: 'New York' },
      phone: '123-456-7890',
      email: 'johndoe@example.com',
    };

    const section = 'Leads';
    const { getByText, getByTestId } = render(
      <QueryClientProvider client={client}>
        <UserDetails route={{ params: { user, section } }} />
      </QueryClientProvider>
    );

    expect(getByText('John Doe')).toBeTruthy();
    expect(getByText('New York')).toBeTruthy();
  });

  it('renders correctly with section "Leads"', () => {
    const user = {
      image: 'user-image-url',
      firstName: 'John',
      lastName: 'Doe',
      username: 'johndoe123',
      address: { city: 'New York' },
      phone: '123-456-7890',
      email: 'johndoe@example.com',
    };

    const section = 'Leads';
    const { getByTestId } = render(
      <UserDetails route={{ params: { user, section } }} />
    );

    expect(getByTestId('national-registry-identification-loader')).toBeTruthy();
    expect(getByTestId('judicial-records-loader')).toBeTruthy();
  });
});
