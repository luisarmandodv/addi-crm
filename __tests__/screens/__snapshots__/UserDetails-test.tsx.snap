// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetails Component Should render snapshot 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "#82ccdd",
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Array [
            Object {
              "marginBottom": 25,
              "paddingHorizontal": 30,
            },
            Object {
              "backgroundColor": "#f0ece2",
              "borderRadius": 10,
              "margin": 8,
              "marginBottom": 10,
              "padding": 10,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "marginLeft": 7,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "rgba(103, 80, 164, 1)",
                  "borderRadius": 40,
                  "height": 80,
                  "width": 80,
                },
                undefined,
              ]
            }
          >
            <Image
              accessibilityIgnoresInvertColors={true}
              source={
                Object {
                  "uri": "user-image-url",
                }
              }
              style={
                Object {
                  "borderRadius": 40,
                  "height": 80,
                  "width": 80,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "marginLeft": 20,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "rgba(28, 27, 31, 1)",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "color": "rgba(28, 27, 31, 0.87)",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 20,
                        "letterSpacing": 0.15,
                        "lineHeight": 30,
                        "marginVertical": 2,
                      },
                      Array [
                        Object {
                          "fontSize": 24,
                          "fontWeight": "bold",
                        },
                        Object {
                          "marginBottom": 5,
                          "marginTop": 15,
                        },
                      ],
                    ],
                  ],
                ]
              }
            >
              John Doe
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "rgba(28, 27, 31, 1)",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "color": "rgba(28, 27, 31, 0.54)",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 12,
                        "letterSpacing": 0.4,
                        "lineHeight": 20,
                        "marginVertical": 2,
                      },
                      Object {
                        "fontSize": 14,
                        "fontWeight": "500",
                        "lineHeight": 14,
                      },
                    ],
                  ],
                ]
              }
            >
              johndoe123
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "marginBottom": 25,
                "paddingHorizontal": 30,
              },
              Object {
                "marginLeft": "-10%",
                "marginTop": "10%",
              },
            ]
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "marginBottom": 16,
              }
            }
          >
            <MockedMaterialCommunityIcons
              color="#777777"
              name="map-marker-radius"
              size={20}
            />
            <Text
              style={
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                  },
                  Object {
                    "writingDirection": "ltr",
                  },
                  Object {
                    "color": "#777777",
                    "marginLeft": 20,
                  },
                ]
              }
            >
              New York
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "marginBottom": 16,
              }
            }
          >
            <MockedMaterialCommunityIcons
              color="#777777"
              name="phone"
              size={20}
            />
            <Text
              style={
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                  },
                  Object {
                    "writingDirection": "ltr",
                  },
                  Object {
                    "color": "#777777",
                    "marginLeft": 20,
                  },
                ]
              }
            >
              123-456-7890
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "marginBottom": 16,
              }
            }
          >
            <MockedMaterialCommunityIcons
              color="#777777"
              name="email"
              size={20}
            />
            <Text
              style={
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                  },
                  Object {
                    "writingDirection": "ltr",
                  },
                  Object {
                    "color": "#777777",
                    "marginLeft": 20,
                  },
                ]
              }
            >
              johndoe@example.com
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "borderBottomColor": "#777777",
              "borderBottomWidth": 1,
              "borderTopColor": "#777777",
              "borderTopWidth": 1,
              "flexDirection": "row",
              "height": 100,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "width": "50%",
                },
                Object {
                  "borderRightColor": "#dddddd",
                  "borderRightWidth": 1,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "rgba(28, 27, 31, 1)",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "color": "rgba(28, 27, 31, 0.54)",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 12,
                        "letterSpacing": 0.4,
                        "lineHeight": 20,
                        "marginVertical": 2,
                      },
                      Object {
                        "marginBottom": 10,
                      },
                    ],
                  ],
                ]
              }
            >
              National Registry Identification
            </Text>
            <ActivityIndicator
              animating={true}
              hidesWhenStopped={true}
              size="small"
              testID="national-registry-identification-loader"
            />
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "width": "50%",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "rgba(28, 27, 31, 1)",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "color": "rgba(28, 27, 31, 0.54)",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 12,
                        "letterSpacing": 0.4,
                        "lineHeight": 20,
                        "marginVertical": 2,
                      },
                      Object {
                        "marginBottom": 10,
                      },
                    ],
                  ],
                ]
              }
            >
              Judicial Records
            </Text>
            <ActivityIndicator
              animating={true}
              hidesWhenStopped={true}
              size="small"
              testID="judicial-records-loader"
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
